bundles/org.openhab.binding.mqtt.generic/src/main/java/org/openhab/binding/mqtt/generic/ChannelState.java

```diff
@@ -352,29 +352,36 @@ public boolean hasSubscribed() {
 
         // Outgoing transformations
         for (ChannelStateTransformation t : transformationsOut) {
-            String transformedValue = t.processValue(mqttCommandValue);
+            String commandString = mqttCommandValue.getMQTTpublishValue(null);
+            String transformedValue = t.processValue(commandString);
             if (transformedValue != null) {
-                mqttCommandValue = transformedValue;
+                Value textValue = new TextValue();
+                textValue.update(new StringType(transformedValue));
```

Comment:
What happens if `cachedValue` is a `NumberValue` and we have a transformation that e.g. returns a character or a string based on the number value? In `update` we do `newValue = new BigDecimal(command.toString());` which will fail.

Reply:
But here we do the `update()` on a new `TextValue` (`textValue`), which should handle the `StringType` just fine.
The result of every transformation stored in a new TextValue. After the transformation the original cachedValue is no longer available, only the new TextValue.

But this also means, that after a transformation, only the format `%s` (plus pre-/postifx) makes any sense.

Reply:
But isn‘t that the same we have now? %d will not work after a transformation. IIRC that was the original request.

Reply:
No, there original request does not contain any tranformation. `%d` does not work even then, because everything is already a string when we start to format. And the Formatter does not try to convert strings to integers.